충남대 BE_이은경 2주차 과제 (2단계)

pkyung opened the pull request:
## 과제 진행
### 인증
- [x] 사용자 도메인 설계
- [x] 사용자 회원 가입 api
- [x] 사용자 로그인 - accessToken

pkyung committed:
Chore: 1단계 코드 이동

pkyung committed:
Chore: 기능 목록 이동

pkyung committed:
Chore: 1주차 피드백 반영 코드 이동

pkyung committed:
Docs: 기능 목록 작성 - 1단계

pkyung committed:
Refactor: final 키워드를 추가하여 autowired 제거

pkyung committed:
Docs: 기능 목록 수정

pkyung committed:
Feat: ValidProductName 어노테이션 생성

ValidProductName 어노테이션을 생성하고 ProductNameValidator를 구현하여
상품 이름의 길이가 15자 이내인지, 특정 특수 문자만 들어있는지, 카카오
라는 단어가 들어있는 지 확인하고, 그에 맞는 응답을 하도록 구현했다.
controller에 ExceptionHandler를 통해 validated로 발생되는
MethodArgumentNotValidException 에러를 잡아서 내부의 defaultMessage를
가져와 400 bad request 와 함꼐 메시지를 반환하도록 구현했다.

pkyung committed:
Refactor: 필요 없는 import 제거 및 코드 정렬

pkyung committed:
Docs: 기능 목록 구현 - 1단계

pkyung committed:
Merge branch 'pkyung' into main

pkyung committed:
Refactor: 기본 Validate 어노테이션으로 수정함

Validator를 제작해서 처리했던 상품 이름 validation을 기본 Validate 어노테이션을 활용하여 구현함

pkyung committed:
Refactor: ExceptionHandler 클래스로 분리

컨트롤러에 예외처리 로직이 있어서 코드가 복잡해졌다. 이를
ProductExceptionHandler 클래스를 만들어 ProductController의 예외를
처리하도록 구현했다.

pkyung committed:
Refactor: Validated를 Valid로 변경

pkyung committed:
Docs: 기능 목록 작성 - 2단계

pkyung committed:
Refactor: 패키지 정리

pkyung committed:
Feat: 사용자 모델 설계

2단계 과제의 요청에 따라 user 객체에는 email과 password를 필드로
설정했다. 그리고 schema에 user 테이블을 생성하는 sql을 작성했다.

pkyung committed:
Feat: 사용자 회원 가입 api

1. userDao : insert와 findByEmail을 만들어 사용자를 데이터베이스에
   삽입하고, 조회하도록 구현했다.
2. UserService : makeUser와 getUser로 사용자 생성, 조회의 비즈니스
   로직을 처리한다. 이 때, 계징이 이미 존재하거나, 존재하지 않는 계정일
   때, 미리 만들어둔 예외를 던진다.
3. UserController : PostMapping과 GetMapping을 통해 사용자 요청을
   처리한다.
4. exception : UserAlreadyExistsException과 UserNotFoundException을
   생성했고, 이를 UserExceptionHandler에서 담당하여 controller에서
   에러 발생 시에 처리한다.
5. UserDto : client와 controller 사이, controller와 service 사이에서
   값을 전달한다.

pkyung committed:
Feat: 사용자 로그인 시에 token 발급

1. JwtUtil : 메서드를 모두 static으로 만들어 다른 클래스에서 의존성
   주입을 하지 않아도 된다. 이 클래스는 JWT 토큰을 생성하고 만료되었는지
   확인하는 등의 메서드가 구현되어 있다.
2. LoginService : 로그인 시, 사용자 정보가 있는 지, 비밀번호가 일치하는
   지 확인하고 맞다면 토큰을 만들어 반환한다. 없다면 사용자를 찾을 수
   없다는 에러를 반환한다. 이 에러는 미리 만들어 둔
   UserExceptionHandler에서 함꼐 잡는다.
3. LoginContorller : 로그인 요청이 들어오면 service를 호출한다.
4. build.gradle : jwt 관련 의존성을 추가했다.

pkyung committed:
Docs: 기능 목록 구현 - 2단계

pkyung committed:
Merge remote-tracking branch 'upstream/pkyung'

# Conflicts:
#	README.md
#	src/main/java/gift/controller/ProductController.java
#	src/main/java/gift/model/Product.java
#	src/main/java/gift/repository/ProductDao.java

ksy90101 commented on src/main/java/gift/controller/LoginController.java:
```diff
@@ -0,0 +1,25 @@
+        this.loginService = loginService;
+    }
+
+    @PostMapping("/token")
```
Login/token이 아니라 login이 어울리는 URL일거 같아요~

pkyung replied:
앗 요구사항에서 login/token으로 요청이 들어와서 이렇게 했습니다.

ksy90101 replied:
요구사항을 다시 한번 보시면 좋을거 같습니다~!
좀더 RESTful하게 URL을 작성해보시면 좋을거 같아요!

ksy90101 commented on src/main/java/gift/controller/UserController.java:
```diff
@@ -0,0 +1,30 @@
+        this.userService = userService;
+    }
+
+    @PostMapping("/user")
```
REST API규칙에선 리소스명을 복수로 씁니다.
users라고 쓰는게 더 어울릴거 같아요!

ksy90101 commented on src/main/java/gift/controller/LoginController.java:
```diff
@@ -0,0 +1,25 @@
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RestController;
+
+@RequestMapping("/login")
```
여기도 실제 api일꺼 같으니 api를 붙이면 어덜까요?

ksy90101 commented on src/main/java/gift/controller/LoginController.java:
```diff
@@ -0,0 +1,25 @@
+
+@RequestMapping("/login")
+@RestController
+public class LoginController {
```
차후 Logout과 같은게 생긴다면 또 컨트롤러를 만들어야 할거 같아요.
좀더 응집도를 가지기 위해서 AuthController라고 인증에 대한 액션들을 모아둔다는 의미로 이름을 지어보면 어떨까요?

ksy90101 commented on src/main/java/gift/dto/UserRequestDto.java:
```diff
@@ -0,0 +1,14 @@
+package gift.dto;
+
+public class UserRequestDto {
```
Reqeust라고 쓸때는 DTO를 굳이 쓸필요는 없다고 생각해요.
Request라는것으로 DTO라는것을 의미하고 잇다고 생각합니다.

ksy90101 commented on src/main/java/gift/exception/ProductExceptionHandler.java:
```diff
@@ -0,0 +1,33 @@
+import org.springframework.web.bind.annotation.RestControllerAdvice;
+
+@RestControllerAdvice(assignableTypes = ProductController.class)
+public class ProductExceptionHandler {
```
👍

ksy90101 commented on src/main/java/gift/jwt/JwtUtil.java:
```diff
@@ -0,0 +1,44 @@
+
+@Component
+public class JwtUtil {
+    private static final Key SECRET_KEY = Keys.hmacShaKeyFor("secret-key-secret-key-secret-key-secret-key".getBytes());
```
해당 키는 중요한 보안상 정보일 확률이 높습니다.

따라서 application.yml에 넣는 편인데요.

한번 @value에 대해서 알아보시면 좋을거 같습니다.

https://rutgo-letsgo.tistory.com/entry/Value%EC%99%80-ConfigurationProperties%EB%A5%BC-%EC%9D%B4%EC%9A%A9%ED%95%B4-yaml-%ED%8C%8C%EC%9D%BC-%EB%A7%A4%ED%95%91%ED%95%98%EA%B8%B0

참고 링크 전달드려요!

ksy90101 commented on src/main/java/gift/jwt/JwtUtil.java:
```diff
@@ -0,0 +1,44 @@
+        return Jwts.builder()
+                .setClaims(claims)
+                .setIssuedAt(new Date(System.currentTimeMillis()))
+                .setExpiration(new Date(System.currentTimeMillis() + 1000 * 60 * 60 * 10))
```
1000 * 60 * 60 * 10

해당 부분을 상수로 뺴서 의미를 전달하면 어떨까요?

ksy90101 commented on src/main/java/gift/model/Product.java:
```diff
@@ -0,0 +1,43 @@
+public class Product {
+    private Long id;
+
+    @NotEmpty(message = "상품 이름이 옳지 않습니다.")
```
MotEmpty가 Length랑 중복이지 않을까요?
한번 테스트 해보시면 좋을거 같아요.

이걸 빼도 빈값이 들어왔을때 문제가 안되는지요~

ksy90101 commented on src/main/java/gift/service/LoginService.java:
```diff
@@ -0,0 +1,27 @@
+
+    public LoginResponseDto login(LoginRequestDto requestDto) {
+        User user = userDao.findByEmail(requestDto.getEmail());
+        if (user == null || !user.getPassword().equals(requestDto.getPassword())) {
```
user.getPassword().equals(requestDto.getPassword())
이런부분들은 도메인에게 직접 메시지를 던지는게 좋습니다.

user.matchPasswrod와 같은 메서드를 만들어 보면 어떨까요?

ksy90101 commented on src/main/resources/schema.sql:
```diff
@@ -3,4 +3,9 @@ CREATE TABLE IF NOT EXISTS products (
+);
+
+CREATE TABLE IF NOT EXISTS users (
+    email VARCHAR(255) PRIMARY KEY,
```
email을 PK로 하신 이유가 따로 있나요?

pkyung replied:
email이 유일한 정보라고 생각했고, jwt 토큰을 생성할 때도 loginId인 email로 만들어서 pk로 지정했습니다.

ksy90101 replied:
https://stackoverflow.com/questions/3804108/use-email-address-as-primary-key

한번 이글을 읽고 다시 한번 고민해보시면 좋을거 같습니다~

ksy90101 reviewed (changes requested):
2단계도 고생 많으셨습니다~
간단한 피드백을 남겼으니 확인해주시면 감사하겠습니다~
궁금한게 있다면 언제든 DM주세요!

pkyung committed:
Refactor: 2단계 멘토님 피드백 반영

1. dto에 request나 response가 들어있다면 dto로 당연시 인식되기 때문에
   dto를 생략 가능하다.
2. LoginController, LoginService를 따로 나누면 loginout 시에도
   LogoutController, LogoutService를 생성해야 하기 때문에
   AuthController, AuthService로 이름을 수정했다.
3. ProductDto의 @NotEmpty는 @Length와 겹쳐 제거했다.
4. JwtUtil의 secretKey는 중요한 정보이기 때문에 application.properties에
   저장한 값을 가져왔다. 하지만 static이기 때문에 Value 어노테이션으로
   한 번에 가져올 수는 없었고, setter를 통해 가져왔다.
5. expiredTime을 상수로 지정했다.

pkyung committed:
Fix: 비밀번호 확인 로직 반대로된 것을 확인 후 수정

pkyung committed:
Refactor: application.properties ignore 파일에 추가

ksy90101 commented on src/main/java/gift/controller/ProductController.java:
```diff
@@ -35,7 +38,7 @@ public ResponseEntity<Product> getProduct(@PathVariable("id") Long id) {
 
     @PutMapping("/products")
-    public ResponseEntity putProduct(@RequestBody @Valid ProductRequestDto requestDto) {
+    public ResponseEntity putProduct(@RequestBody @Valid ProductRequest requestDto) {
```
변수명도 같이 변경하면 어떨까요?

ksy90101 commented on src/main/java/gift/dto/LoginRequest.java:
```diff
@@ -0,0 +1,14 @@
+package gift.dto;
+
+public class LoginRequest {
+    private String email;
```
Email에 대한 Validate를 해보면 어떨까요?
잘못된 Email 형식이 들어온다면요~
1단계에서 Validate를 배웠으니 쉽게 가능할거라고 생각이 들어요!

ksy90101 commented on src/main/java/gift/exception/UserAlreadyExistsException.java:
```diff
@@ -0,0 +1,7 @@
+package gift.exception;
+
+public class UserAlreadyExistsException extends RuntimeException {
```
차후 excpetion이 많아 진다면 보기가 어려울거 같기도 해요.
패키지 구조를 한번 도메인별대로 가져가면 어떨까요?
user와 product를 패키지를 만들어서 관리하는거죠.
차후 그러면 찾기도 편할거 같아요~

ksy90101 commented on src/main/java/gift/jwt/JwtUtil.java:
```diff
@@ -0,0 +1,52 @@
+    private static String SECRET_KEY;
+    private static final int EXPIRED_TIME = 1000 * 60 * 60 * 10;
+
+    @Value("${jwt.secret-key}")
```
👍

ksy90101 commented on src/main/java/gift/service/AuthService.java:
```diff
@@ -0,0 +1,36 @@
+    }
+
+    private boolean userMatchPassword(User user, String inputPassword) {
+        if (user.getPassword().equals(inputPassword)) {
```
어.. 이런말씀이 아니라 user 객체에 메서드를 만들라는 이야기였어요~
도메인 로직으로요!

ksy90101 commented on src/main/java/gift/repository/UserDao.java:
```diff
@@ -0,0 +1,40 @@
+        jdbcTemplate.update(sql, user.getEmail(), user.getPassword());
+    }
+
+    public User findByEmail(String email) {
```
한번 심화적인 내용으로 Optional을 써보면 어떨까요?
좀더 안정적인 코드가 될거 같기도 해요~

ksy90101 reviewed (changes requested):
고생하셨습니다~
좀더 심화적인 내용들을 작성해봤어요!
한번 고민해보시고 궁금한게 있다면 언제든 DM주세요!

pkyung committed:
Refactor: 2단계 요구사항 명세 수정

1. 기존의 user를 생성하는 makeUser 메서드를 Auth로 옮겨 register로 수정
2. accessToken이던 이름을 token으로 변경

pkyung committed:
Fix: 비밀번호 틀릴 시 forbidden 반환으로 수정

pkyung committed:
Refactor: exception 관리 패키지 별로 수정함

ksy90101 reviewed (changes requested):
남긴 추가 리뷰 확인해주세요~

pkyung committed:
Refactor: 멘토님 피드백 반영

1. requestDto -> request 로 변수명 수정
2. LoginRequest와 UserReqeust 의 email validate 추가
3. email pk id로 auto increment
4. User 객체에 matchPassword 추가
5. 한 개만 조회 시에 optional 객체 사용

pkyung committed:
Refactor: 필요없는 구절 제거

ksy90101 commented on src/main/resources/schema.sql:
```diff
@@ -3,4 +3,10 @@ CREATE TABLE IF NOT EXISTS products (
+
+CREATE TABLE IF NOT EXISTS users (
+    id BIGINT PRIMARY KEY AUTO_INCREMENT,
+    email VARCHAR(255) ,
```
이메일은 중복이 되면 안되도록 해둔거 같아요~
제약사항에서도 unique index를 써면 어떨가요!?

ksy90101 commented on src/main/java/gift/model/User.java:
```diff
@@ -0,0 +1,28 @@
+    }
+
+    public boolean matchPassword(String password) {
+        if (this.password.equals(password)) {
+            return true;
+        }
+        return false;
```
```suggestion
    return this.password.equals(password)
```
이런식으로 already return해도 될거 같아요!

ksy90101 reviewed (approved):
2단계도 고생 많으셨습니다~
간단한 피드백을 남기긴 했는데 3단계에서 같이 고민해보시면 좋을거 같습니다~
언제든 질문이 있다면 DM주세요!
